File: app/src/main/java/com/example/starwars/repository/FilmsRepository.java
package com.example.starwars.repository;

import static com.example.starwars.activity.FilmListActivity.TAG;

import android.os.AsyncTask;
import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.starwars.database.StarWarsDatabase;
import com.example.starwars.model.Film;
import com.example.starwars.model.Films;
import com.example.starwars.retrofit.RetrofitClient;
import com.example.starwars.retrofit.StarAPI;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.annotations.NonNull;
import io.reactivex.rxjava3.observers.DisposableSingleObserver;
import io.reactivex.rxjava3.schedulers.Schedulers;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class FilmsRepository {

    public static final String API_KEY = "430dc4bcb90f3bd8e2616b75a712749c";

    private static FilmsRepository instance;
    private StarAPI mStarAPI;
    private StarWarsDatabase mDatabase;
    private MutableLiveData<List<Film>> mFilmList = new MutableLiveData<>();
    private MutableLiveData<Boolean> mLoadingIndicator = new MutableLiveData<>();
    private MutableLiveData<Film> mFilm = new MutableLiveData<>();

    int currentPage = 0;
    int maxPage = 1;

    private FilmsRepository() {
        Retrofit retrofit = RetrofitClient.getInstance();
        mStarAPI = retrofit.create(StarAPI.class);
        mDatabase = StarWarsDatabase.getInstance();
    }

    public static FilmsRepository getInstance() {
        if (instance == null) {
            instance = new FilmsRepository();
        }
        return instance;
    }

    public MutableLiveData<List<Film>> getFilmListFromAPI() {
        if (currentPage < maxPage) {
            currentPage++;
            mLoadingIndicator.postValue(true);
            mStarAPI.getBaseFilms(API_KEY, currentPage)
                    .enqueue(new Callback<Films>() {
                        @Override
                        public void onResponse(Call<Films> call, Response<Films> response) {
                            insertFilms(response.body().getFilms());
                            List<Film> currentFilms = mFilmList.getValue();
                            if (currentFilms == null)
                                currentFilms = new ArrayList<>();
                            currentFilms.addAll(response.body().getFilms());
                            mLoadingIndicator.postValue(false);
                            mFilmList.postValue(currentFilms);
                            maxPage = response.body().getPages();
                        }

                        @Override
                        public void onFailure(Call<Films> call, Throwable t) {
                            currentPage = 0;
                            getFilmListFromDatabase();
                        }
                    });
        }
        return mFilmList;
    }

    private void insertFilms(List<Film> films) {
        new AsyncTask<List<Film>, Void, Void>() {

            @Override
            protected Void doInBackground(List<Film>... film) {
                for (int i = 0; i < film[0].size(); i++) {
                    mDatabase.mFilmsDao().insert(film[0].get(i));
                }
                return null;
            }
        }.execute(films);
    }

    private void updateFilm(Film film) {
        new AsyncTask<Film, Void, Void>() {

            @Override
            protected Void doInBackground(Film... films) {
                mDatabase.mFilmsDao().update(film);
                return null;
            }
        }.execute(film);
    }

    public void getFilmListFromDatabase() {
        mDatabase.mFilmsDao().getFilms()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new DisposableSingleObserver<List<Film>>() {
                    @Override
                    public void onSuccess(@NonNull List<Film> films) {
                        mFilmList.postValue(films);
                        mLoadingIndicator.postValue(false);
                    }

                    @Override
                    public void onError(@NonNull Throwable e) {
                        mFilmList.postValue(null);
                        mLoadingIndicator.postValue(false);
                    }
                });
    }

    public void getFilmFromDatabase(int id) {
        mDatabase.mFilmsDao().getFilm(id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new DisposableSingleObserver<Film>() {
                    @Override
                    public void onSuccess(@NonNull Film film) {
                        mFilm.postValue(film);
                        mLoadingIndicator.postValue(false);
                    }

                    @Override
                    public void onError(@NonNull Throwable e) {
                        mFilm.postValue(null);
                        mLoadingIndicator.postValue(false);
                    }
                });
    }

    public MutableLiveData<Boolean> getLoadingIndicator() {
        return mLoadingIndicator;
    }

    public LiveData<Film> getFilmFromAPI(int id) {
        mLoadingIndicator.postValue(true);
        mStarAPI.getFilm(id, API_KEY)
                .enqueue(new Callback<Film>() {
                    @Override
                    public void onResponse(Call<Film> call, Response<Film> response) {
                        updateFilm(response.body());
                        mFilm.postValue(response.body());
                        Log.d(TAG, "onResponse: ");
                        mLoadingIndicator.postValue(false);
                    }

                    @Override
                    public void onFailure(Call<Film> call, Throwable t) {
                        getFilmFromDatabase(id);
                        Log.d(TAG, "onFailure: ");
                    }
                });
        return mFilm;
    }

}

File: app/src/main/java/com/example/starwars/repository/FilmsRepository.kt
package com.example.starwars.repository

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.example.starwars.activity.FilmListActivity.Companion.TAG
import com.example.starwars.database.StarWarsDatabase
import com.example.starwars.model.Film
import com.example.starwars.retrofit.StarAPI
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.schedulers.Schedulers
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class FilmsRepository @Inject constructor(private val mStarAPI: StarAPI, private val database: StarWarsDatabase) {

    private val mFilmList = MutableLiveData<List<Film>?>()
    val loadingIndicator = MutableLiveData<Boolean>()
    private val mFilm = MutableLiveData<Film?>()

    private var maxPage = 1


    private fun insertFilms(films: List<Film>) {
        Observable.fromCallable { this.database.mFilmsDao().insertAll(films) }
            .subscribeOn(Schedulers.io())
            .subscribe()
    //  Reactive variant:
    //
//        Observable.fromIterable(films)
//            .subscribeOn(Schedulers.io())
//            .subscribe{ film -> mDatabase.mFilmsDao().insert(film) }
    }

    private fun updateFilm(film: Film) {
        Observable.fromCallable {
            -> this.database.mFilmsDao().update(film)
            Log.d(TAG, "updateFilm: " + film.title + " " + film.crews)
        }
            .subscribeOn(Schedulers.io())
            .subscribe()
    }

    fun filmListFromAPI(currentPage: Int): MutableLiveData<List<Film>?> {
            if (currentPage <= maxPage) {
                loadingIndicator.postValue(true)
                mStarAPI.getBaseFilms(currentPage)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .doOnSuccess { data -> data.films?.let { insertFilms(it) } }
                    .subscribe({ data ->
                        val currentFilms: MutableList<Film>
                        if (mFilmList.value != null) currentFilms = mFilmList.value as MutableList
                        else currentFilms = ArrayList()
                        data.films?.let { currentFilms.addAll(it) }
                        loadingIndicator.postValue(false)
                        mFilmList.postValue(currentFilms)
                        maxPage = data.pages
                    }, {
                        Log.d(TAG, "filmListFromAPI got error: " + it.message)
                    })
            }
        return mFilmList
    }

    fun getFilmListFromDatabase(): MutableLiveData<List<Film>?> {
            this.database.mFilmsDao().films
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                        mFilmList.postValue(it)
                        loadingIndicator.postValue(false)
                }, {
                    mFilmList.postValue(null)
                    loadingIndicator.postValue(false)
                })
        return mFilmList
    }

    fun getFilmFromDatabase(id: Int): LiveData<Film?>  {
        this.database.mFilmsDao().getFilm(id)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({
                mFilm.postValue(it)
                Log.d(TAG, "getFilmFromDatabase: " + it.title)
                Log.d(TAG, "and crew: " + it.crews)
                loadingIndicator.postValue(false)
            }, {
                mFilm.postValue(null)
                Log.d(TAG, "getFilmFromDatabase: Error!!")
                loadingIndicator.postValue(false)
            })
        return mFilm
    }

    fun getFilmFromAPI(id: Int): LiveData<Film?> {
        loadingIndicator.postValue(true)
        mStarAPI.getFilm(id)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe( {
                updateFilm(it)
                Log.d(TAG, "getFilmFromAPI: " + it.title + " " + it.crews + " and updated")
                mFilm.postValue(it)
                loadingIndicator.postValue(false)
            }, {
                mFilm.postValue(null)
                loadingIndicator.postValue(false)
            })
        return mFilm
    }

}